dev/DynamicDependency/API/MddDetourPackageGraph.cpp

```diff
@@ -120,18 +142,35 @@ HRESULT WINAPI MddDetourPackageGraphInitialize() noexcept
     // Detour package graph APIs to our implementation
     FAIL_FAST_IF_WIN32_ERROR(DetourUpdateThread(GetCurrentThread()));
     FAIL_FAST_IF_WIN32_ERROR(DetourAttach(&(PVOID&)TrueGetCurrentPackageInfo, DynamicGetCurrentPackageInfo));
-    FAIL_FAST_IF_WIN32_ERROR(DetourAttach(&(PVOID&)TrueGetCurrentPackageInfo2, DynamicGetCurrentPackageInfo2));
-    if (IsWindows10_20H1OrGreater())
+    //
+    // NOTE: GetCurrentPackageInfo2 requires >=19H1
+    // NOTE: GetCurrentPackageInfo3 requires >=20H1
+    // Support down to RS5
+    if (IsWindows10_19H1OrGreater())
     {
-        HMODULE dllKernelbase{ LoadLibraryExW(L"kernelbase.dll", nullptr, 0) };
-        FAIL_FAST_HR_IF_NULL(HRESULT_FROM_WIN32(GetLastError()), dllKernelbase);
+        HMODULE dllApisetAppmodelRuntime_1_3{ LoadLibraryExW(L"api-ms-win-appmodel-runtime-l1-1-3.dll", nullptr, 0) };
+        FAIL_FAST_HR_IF_NULL(HRESULT_FROM_WIN32(GetLastError()), dllApisetAppmodelRuntime_1_3);
 
-        auto dllGetCurrentPackageInfo3{ reinterpret_cast<GetCurrentPackageInfo3Function>(GetProcAddress(dllKernelbase, "GetCurrentPackageInfo3")) };
-        FAIL_FAST_HR_IF_NULL(HRESULT_FROM_WIN32(GetLastError()), dllGetCurrentPackageInfo3);
+        auto dllGetCurrentPackageInfo2{ reinterpret_cast<GetCurrentPackageInfo2Function>(GetProcAddress(dllApisetAppmodelRuntime_1_3, "GetCurrentPackageInfo2")) };
+        FAIL_FAST_HR_IF_NULL(HRESULT_FROM_WIN32(GetLastError()), dllGetCurrentPackageInfo2);
+        TrueGetCurrentPackageInfo2 = dllGetCurrentPackageInfo2;
+        FAIL_FAST_IF_WIN32_ERROR(DetourAttach(&(PVOID&)TrueGetCurrentPackageInfo2, DynamicGetCurrentPackageInfo2));
 
-        TrueGetCurrentPackageInfo3 = dllGetCurrentPackageInfo3;
+        // Also grab GetPackageInfo2 if available (see MddGetPackageInfo1Or2() for details)
+        auto dllGetPackageInfo2{ reinterpret_cast<GetPackageInfo2Function>(GetProcAddress(dllApisetAppmodelRuntime_1_3, "GetPackageInfo2")) };
+        FAIL_FAST_HR_IF_NULL(HRESULT_FROM_WIN32(GetLastError()), dllGetPackageInfo2);
+        TrueGetPackageInfo2 = dllGetPackageInfo2;
 
-        FAIL_FAST_IF_WIN32_ERROR(DetourAttach(&(PVOID&)TrueGetCurrentPackageInfo3, DynamicGetCurrentPackageInfo3));
+        if (IsWindows10_20H1OrGreater())
+        {
+            HMODULE dllApisetAppmodelRuntimeInternal_1_6{ LoadLibraryExW(L"api-ms-win-appmodel-runtime-internal-l1-1-6.dll", nullptr, 0) };
```

Comment:
Note: Does this work on Windows 11 22000+ where apiset is `api-ms-win-appmodel-runtime-internal-l1-1-9.dll`?

Reply:
Yes

Naming scheme is blahblah-l1-major-minor, so that's version 1.9 of that APIset. APIsets follow a strong breaking-change-requires-major-change policy so if 1-9 is present so too is 1-0 thru 1-8. 1-9 _adds_ symbols over 1-8; it's generally problematic to delete or change existing symbols (not impossible but not usually done, for obvious functional reasons as well as technical). When that needs to happen it's time to make 2-0.

That doesn't mean everything in 1-9 is functional (if it's internal). If, say, 1-5 added DoFoo(), DoFoo wasn't public and it's not obsolete (e.g. everything that used it doesn't anymore let alone if it's problematic now) then we can't remove DoFoo but we can gut the implementation e.g. `{ return E_NOTIMPL; }`.

Reply:
Cool, just wanted to double check apisets we're versioned as such. Wasn't sure. Lgtm.